BTrees/tests/common.py

```diff
@@ -2313,6 +2313,8 @@ def testUnion(self):
                 C = self.union(A, B)
                 self.assertTrue(not hasattr(C, "values"))
                 self.assertEqual(list(C), self._union(A, B))
+                if PY3 or A.__class__ == B.__class__:
+                    self.assertEqual(list(C), list(A | B))
```

Comment:
```suggestion
                    self.assertEqual(set(A) | set(B), set(A | B))
```

This seems like a more fair, direct comparison. Using `C` is relying on the results of `self.union`, which, I think, is not really what we're trying to test for here (and likewise in the rest of the methods); the fact that it's used internally is an implementation detail.

Also, perhaps that loosens the restrictions about needing PY3 and/or equal classes? In any case, if something is throwing a `TypeError`, it would be nice to have a test case showing that.

Reply:
Your assertion suggestion sounds good indeed.

Actually I am not really sure what is going on with python2 here. Using those operators between different kinds of datastructures, say a Set and a TreeSet, sometimes result in a `TypeError: unsupported operand type(s)`, and sometimes not.

Reply:
I could finally tighten the tests.
It seems that in python2, using those operators on two different C classes objects raises a TypeError. Any other configuration seems to work.

Reply:
Were you able to debug to the specific C call that’s failing? Is it before or after our code takes control?

Maybe we just need appropriate `__r*__` definitions?

I *think* my preference for the operators would be to be fairly loose in what they accept, especially if it smooths over any 2/3 differences.  Maybe that entails calling the class’s constructor on the argument when needed (for incompatible types, that probably results in a more specific exception)? Those who wish a very strict, fast interpretation can continue to use the existing functions. 

Of course it’s always easier to loosen restrictions rather than add them in the future. My main motivation for these new operators is to try to get the 2/3 semantics as much the same as possible. With `None` we initially went the strict Py3 way everywhere then reverted to the looser Py2 semantics. Here it seems good to go with the looser Py3 version everywhere from the start.

Reply:
I think python2 type checks are related to [this flag](https://docs.python.org/2.7/c-api/typeobj.html#Py_TPFLAGS_CHECKTYPES).

Reply:
Nice discovery!

I'd like to suggest that in the future you try to avoid force-pushing to PRs that are being reviewed. Reviewers get a notification with the same commit message that they've gotten before. When they go to github to see the changes, the github review tools are crippled: they have to review the entire changeset again and try to remember what might be different this time vs the last time.

Moreover, I would argue that the history here, together with good commit messages, would have been extremely valuable. History would have been able to answer the question "Why is Py_TPFLAGS_CHECKTYPES set? What's the purpose?". Now, it's just there and we have to guess what part of the new tests failed without it.

Reply:
Ok I see. Actually I have never really been this side of the patch so I am not familiar with the tools github provides. I will avoid rebase in those situations in the future.